File: cSharp/Classes/PlayerMediator.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace lean_pactheman_client {
    public static class PlayerMediator {
        private static Player _playerInstance { get; set; }
        public static EventWaitHandle WaitRewardHandle = new AutoResetEvent(false);

        public static void SetPlayer(Player p) => _playerInstance = p;
        /// <summary>
        /// Blocks the lean client to receive an immediate update of the player state which can be used afterwards via <c>GameState.Instance</c>.
        /// </summary>
        /// <param name="updateVelocity">The Velocity to apply for the player state update</param>
        /// <returns><c>bool</c>: Wheter update was successful or not</returns>
        public static async Task<bool> ReceivePlayerStateUpdate(Velocity updateVelocity) {
            try {
                _playerInstance.UpdateState(updateVelocity.Normalize());
                await _playerInstance.SendState();
                return WaitRewardHandle.WaitOne();
            } catch (Exception ex) {
                Console.WriteLine(ex.ToString());
                return false;
            }
        }
    }
}

File: dart/lib/classes/player_mediator.dart
import 'package:lean_dart_client/classes/player.dart';
import 'package:lean_dart_client/classes/velocity.dart';
import 'package:lean_dart_client/classes/game_state.dart';

class PlayerMediator {
  static Player _playerInstance;

  static void setPlayer(Player p) => _playerInstance = p;

  static Future<bool> receivePlayerStateUpdate(Velocity update) async {
    try {
      _playerInstance.updateState(update);
      _playerInstance.sendState();
      return await GameState.instance.rewardAvailableStream.first;
    } catch (ex) {
      print(ex.message);
      return false;
    }
  }
}
